File: matlab/matrix-free-stacked/stokes_flow/bfun.m
function w = bfun(u, xy, xyp, mv, bxe, bye)

x = xy(:, 1);
xp = xyp(:, 1);
nvtx = length(x);
nu = 2 * nvtx;
np = 3 * length(xp);
nel = length(mv(:, 1));
mp = [[1:3:3*nel]',[2:3:3*nel]',[3:3:3*nel]'];

ux = u(1:nu/2);
uy = u(nu/2+1:nu);

bxes = squeeze(bxe(1,:,:));
byes = squeeze(bye(1,:,:));

w = zeros(nu+np,1);

for e = 1:nel
    ind9 = mv(e,:);
    wex = bxes * ux(ind9); 
    wey = byes * uy(ind9); 
    
    ind3 = mp(e,:) + nu; 
    w(ind3) = w(ind3) + wex; 
    w(ind3) = w(ind3) + wey;
end

end

File: BrinkmanStokes/src/matrix-free-stacked/stokes_flow/bfun.jl
function bfun(u, kparams)

  xy = kparams["xy"]; xyp = kparams["xyp"]; mv = kparams["mv"]; bound = kparams["bound"];
  bxe = kparams["bxe"]; bye = kparams["bye"]

  # get variables
  nvtx = length(xy[:, 1]); nu = 2nvtx; np = 3length(xyp[:, 1])
  nel = length(mv[:, 1])
  bxes = squeeze(bxe[1, :, :], 1)
  byes = squeeze(bye[1, :, :], 1)
  mp = [[1:3:3nel] [2:3:3nel] [3:3:3nel]]
  w = zeros(nu+np)
  # w = SharedArray(Float64, (nvtx, 1), pids = workers())

  ux = u[1:nvtx]
  uy = u[nvtx + 1:nu]

  # @parallel for e = 1:nel
  for e = 1:nel
      ind9 = vec(mv[e, :]')
      ind3 = vec(mp[e, :]')
      w[ind3] += bxes * ux[ind9]
      w[ind3] += byes * uy[ind9]
  end

  vec(w)
end
